components/raftstore-v2/src/operation/command/admin/commit_merge.rs

```diff
@@ -0,0 +1,927 @@
+    ) -> Result<u64> {
+        if self.storage().has_dirty_data() {
+            return Err(box_err!(
+                "{:?} target peer has dirty data, try again later",
+                self.logger.list()
+            ));
+        }
+        // Match v1::check_merge_proposal.
+        let source_region = req.get_admin_request().get_commit_merge().get_source();
+        let region = self.region();
+        if !util::is_sibling_regions(source_region, region) {
+            return Err(box_err!(
+                "{:?} and {:?} should be sibling",
+                source_region,
+                region
+            ));
+        }
+        if !region_on_same_stores(source_region, region) {
+            return Err(box_err!(
+                "peers not matched: {:?} {:?}",
+                source_region,
+                region
+            ));
+        }
+        let mut proposal_ctx = ProposalContext::empty();
+        proposal_ctx.insert(ProposalContext::COMMIT_MERGE);
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, proposal_ctx.to_vec())
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    // Match v1::exec_commit_merge.
+    pub async fn apply_commit_merge(
+        &mut self,
+        req: &AdminRequest,
+        index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        PEER_ADMIN_CMD_COUNTER.commit_merge.all.inc();
+
+        self.flush();
+
+        // Note: doesn't validate region state from kvdb any more.
+        let merge = req.get_commit_merge();
+        let source_region = merge.get_source();
+        let (tx, rx) = oneshot::channel();
+        self.res_reporter().send(PeerMsg::CatchUpLogs(CatchUpLogs {
+            target_region_id: self.region_id(),
+            merge: merge.clone(),
+            tx,
+        }));
+        let source_ready = rx.await.unwrap_or_else(|_| {
+            // TODO: handle this gracefully.
+            slog_panic!(self.logger, "source peer is missing");
+        });
+
+        info!(
+            self.logger,
+            "execute CommitMerge";
+            "commit" => merge.get_commit(),
+            "entries" => merge.get_entries().len(),
+            "index" => index,
+            "source_region" => ?source_region
+        );
+
+        let mut region = self.region().clone();
+        // Use a max value so that pd can ensure overlapped region has a priority.
+        let version = cmp::max(
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+
+        let source_tablet: EK = match source_ready.tablet.downcast() {
```

Comment:
I was thinking about making source peer create a checkpoint instead of passing db handles around. And unlike v1, v2 can skip calling source peer if checkpoint exists. Because there is no race at all in v2.

If we are going to introduce multi-range raft later, this implementation can also be reused.

Reply:
The purpose is to reduce number of db opening. Right now, it only opens once (after rename to final path), to introduce a checkpoint, it will open twice. I don't know how long it takes for split to open a checkpoint right now. With that information we can decide if it's worth optimizing.

Reply:
Why it needs to be twice? For example, we can merge the two manifest from both checkpoints, rename it and then open.

Reply:
Manipulating manifest without opening a DB is unthinkable. The code to recover files and modify manifest are all coupled with DB. A better choice would be to avoid loading table footer for intermediate DBs.